lib/src/piece/piece.dart

```diff
@@ -210,7 +208,26 @@ class TextPiece extends Piece {
   }
 
   @override
-  String toString() => '`${_lines.join('¬')}`${_containsNewline ? '!' : ''}';
+  String toString() => '`${_lines.join('¬')}`';
+}
+
+/// A single line of text within a [TextPiece].
+class _Line {
+  String _text = '';
```

Comment:
[nit] I think it's overkill to use private fields in a private class. The rest of the library reads some of these directly, so making them private here does not indicate that they are only used within the class definition.

Reply:
I've gone back and forth on this. I used to just make everything public if the class was private (and non-escaping).

But over time, I've appreciated the "unused declaration" warnings more and more and you only get those for private members. So I tend to use private for anything I can that way the analyzer helps me clean up dead code better.